로그인 정보확인 API 연동 및 회원가입 폼 이메일, 닉네임 중복체크 기능 구현

wooojini opened the pull request:
## 📑 제목

#75 #119 로그인 정보확인 API 연동 및 회원가입 폼 이메일, 닉네임 중복체크 기능 구현

## 💬 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 로그인 정보확인 API 연동 작업
   - 최초 HeaderAuthSection 렌더링시 로그인 정보확인 쿼리를 통해 캐시에 로그인 상태 저장
   - 캐시의 로그인 여부에 따라 로그인 또는 유저프로필 정보가 렌더링 되도록 구현
   - 로그인이 안되어있을 경우, 로그인 메뉴에서 로그인 성공시 lazyQuery를 날려서 캐시에 로그인 상태저장 => HeaderAuthSection 리렌더링
   ![ezgif com-gif-maker (40)](https://user-images.githubusercontent.com/32856129/118654085-263f9780-b823-11eb-887a-2e8aa082905b.gif)
- [x] 클라이언트 라우팅 작업
   - 한표 헤더 메뉴에서 로그인이 필요한 메뉴는 캐시에 저장된 로그인 정보에 따라 접근이 가능하도록 수정
   - 로그인이 필요한 경우, 메인페이지로 이동 후 로그인 모달 창이 보여지도록 구현
   ![ezgif com-gif-maker (41)](https://user-images.githubusercontent.com/32856129/118654316-66067f00-b823-11eb-8708-2f46a92caaff.gif)
   ![ezgif com-gif-maker (42)](https://user-images.githubusercontent.com/32856129/118654517-9cdc9500-b823-11eb-83ed-4162b18dd6fd.gif)
- [x] 회원가입 폼 이메일, 닉네임 중복체크 기능 구현
   ![ezgif com-gif-maker (44)](https://user-images.githubusercontent.com/32856129/118655652-b6321100-b824-11eb-86b9-87c3663022f1.gif)

- [x] 기타 작업사항
   - SnackbarStore 관련 로직 추상화
   - ModalStore 관련 로직 추상화 
   - 이메일, 닉네임 중복체크 쿼리 정의에 따른 쿼리 타입 생성

## 🚧 PR 특이 사항


> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- **로그아웃 기능은 서버 API가 아직 구현되지 않아서 미구현입니다.** 
- **회원가입 폼에서 이메일, 닉네임 중복체크를 수행할 때 의도와는 다른 동작이 수행되고 있습니다.**
   - 위에 중복체크를 수행하는 gif를 보시면 눈치채셨을 수 도 있는데, 최초 중복체크 버튼을 클릭하고 이후에는 중복체크 버튼을 클릭하지 않아도 디바운싱 유효성 검사 이후에 자동으로 쿼리를 날려서 중복체크를 수행합니다...?!
   - 사실 UX를 고려하면 더 좋은 동작이 수행되고 있지만, 이유를 모르겠네용! useLazyQuery를 이용해서 중복체크 버튼을 클릭할 때만 중복체크 쿼리를 날리는데 현재는 디바운싱 이후, 리렌더링 시 자동으로 쿼리를 날립니다.
   - useLazyQuery 공식문서를 참조해도 필요한 순간에 쿼리함수를 실행해서 쿼리를 날린다고 되어있는데 이유를 모르겠네요!
- **snackbarStore와 modalStore 관련 로직들을 해당 스토어의 메소드로 추상화 했습니다.**
   - 이전에 잠시 이야기 나눈데로, store 관련 로직들이 store 내부에 추상화가 덜 되어있어서 store에 접근하면 관련 type들도 import하여 메소드를 호출한다던지 번거로운 코드가 많이 존재했습니다.
   - store 관련 로직들을 내부적으로 메소드로 추상화하여 외부에서는 store만 import 후, 메소드만 호출하면 동작하도록 리팩토링 후, 관련 코드들을 수정했습니다 :)
   - 추후에 snackbarStore, modalStore  이외에도 store 코드들을 내부적으로 추상화하는 작업을 진행하면 좋을 것 같네요!!
- **현재 gif는 마이페이지가 헤더 메뉴에 있는 상태이지만 이전에 이야기 나눈데로 마이페이지 메뉴는 제거할 예정입니다 :)**

wooojini committed:
refactor: HeaderAuthSection 컴포넌트 네이밍 변경

- HeaderAuthSection 컴포넌트 네이밍 및 구조 변경

wooojini committed:
refactor: UserProfile 컴포넌트 props 추가

- UserProfile 컴포넌트 렌더링에 필요한 닉네임과 전공 정보를 props로
  전달받도록 수정

wooojini committed:
feat: myMemberInfo graphQL 쿼리 추가

- 로그인 정보를 요청하는 쿼리 추가

wooojini committed:
chore: graphQL 스키마 타입 추가

- myMemberInfo에 대한 graphQL 스키마 타입 추가

wooojini committed:
refactor: 로그인 여부에 따른 프로필 및 로그인 메뉴 렌더링 로직 추가

wooojini committed:
feat: 로그인 모달 폼 로그인 성공시 로그인 정보 캐시에 저장하는 로직 추가

wooojini committed:
feat: AlertSnackbar 헤더 네비게이션 실패 메세지 추가

- 헤더 네비게이션 메뉴 클릭 시, 로그인이 안된 경우 보여줄 메세지 추가

wooojini committed:
refactor: HeaderMenu 컴포넌트 외부에서 네비게이션 로직을 주입받도록 수정

- 로그인 여부에 따라 외부에서 네비게이션 로직을 적용할 수 있도록 코드
  수정

wooojini committed:
feat: HeaderNavSection 컴포넌트 제작

- HeaderMenu에 대한 비즈니스 로직을 정의할 HeaderNavSection 컴포넌트
  제작

wooojini committed:
refactor: SnackbarStore 메세지 메소드 추상화

- snackbarStore에 접근시 별도의 SnackbarType을 참조할 필요없이 상황에
  따른 메세지를 호출할 수 있도록 메소드 정의

wooojini committed:
refactor: SnackbarStore 메소드 추상화로 인한 관련 코드 수정

- SnackbarStore 메소드 추상화로 인한 store 접근 코드 수정

wooojini committed:
refactor: ModalStore 메소드 추상화

- Modal과 관련된 로직들을 ModalStore에 메소드를 정의하여 추상화

wooojini committed:
refactor: ModalStore 메소드 추상화로 인한 관련 코드 수정

- ModalStore 메소드 추상화로 인한 store 접근 코드 수정

wooojini committed:
feat: 이메일 중복체크 쿼리 정의

- 회원가입 시 이메일 중복체크 확인을 위한 쿼리 정의

wooojini committed:
feat: 닉네임 중복체크 쿼리 정의

- 회원가입시 닉네임 중복체크 쿼리 정의

wooojini committed:
chore: 이메일, 닉네임 중복체크 쿼리 타입 생성

wooojini committed:
feat: 회원가입 폼 이메일, 닉네임 중복체크 기능 구현

- 이메일, 닉네임 중복체크 버튼 클릭시 쿼리를 통해 중복체크를 진행할 수
  있도록 구현
- 중복체크 여부에 따른 회원가입 버튼 활성화 및 helperText 표시 로직 추가

wooojini committed:
refactor: 헤더 메뉴에서 마이페이지 메뉴 제거

- 마이페이지 메뉴는 헤더에서 보여주는게 불필요하므로 로그인 이후, 프로필
  메뉴에서 접근 가능하도록 수정

jinhyukoo commented on src/components/UI/molecules/HeaderNavSection/HeaderNavSection.tsx:
```diff
@@ -0,0 +1,53 @@
+  const checkAuthState = () => {
+    const meberInfo = client.readQuery({ query: MY_MEMBER_INFO });
+
+    return !!meberInfo;
```
!! => 요게 어떤 문법이었죠 우진님...? (분명히 예전에 배웠는데 까먹었네용...)

이것 덕에 readQuery에서 불러올 정보가 없어도 에러가 뜨지 않는 건가용?

wooojini replied:
@jinhyukoo 

!는 단일 연산자로 원래 용도는 boolean 값의 결과를 반대로 평가해주는 연산자입니다!
```
console.log(!true) //false;
``` 

JS에서는 연산자를 사용한 표현식들을 평가할 때 자동 형 변환을 수행합니다! 예를 들어 +연산자의 경우 원래는 산술 연산을 위한 연산자입니다. JS 엔진에서 +연산자는 산술 연산에 쓰이기 때문에 피연산자들을 number 값으로 자동으로 형변환을 시키고 연산을 수행합니다. 이러한 동작 방식 때문에 +연산자를 단일 연산자로 사용하면 아래와 같이 형변환이 일어납니다!
```
console.log(+false) //0
``` 

!연산자도 마찬가지로 boolean 값을 대상으로 수행하는 연산자이기 때문에 피연산자를 boolean 값으로 형변환을 합니다!
```
console.log(!''); //true => ''은 falsy한(false로 평가되는) 값이므로 false로 평가되고, !연산자를 수행해 true로 바뀜
```

위 원리를 이용해서 !!처럼 두 번 사용하면 피연산자를 boolean 값으로 형 변환을 수행할 수 있습니다. 
```
console.log(!!''); //false => !연산자를 수행해 true로 바뀐 값을 다시 !연산자를 수행해 false로 바꿈
```

위 코드는 로그인 정보가 캐시에 저장되어 있는지 없는지를 boolean 값으로 형 변환을 수행하는 코드입니다 :)

jinhyukoo commented on src/components/UI/molecules/HeaderNavSection/HeaderNavSection.tsx:
```diff
@@ -0,0 +1,53 @@
+  const onHeaderMenuClickListener = (link: string) => {
+    if (link !== MAIN_PAGE_LINK && !checkAuthState()) {
+      snackbarStore.showNavFailedMsg();
+      history.push(MAIN_PAGE_LINK);
```
Link에서 history.push로 방법을 바꾸신 이유는 조건을 걸어서 로그인 여부를 확인하기 위함일까요? 아니면 다른 이유가 있으신지 궁금합니당!

wooojini replied:
@jinhyukoo 

헤더 메뉴에서는 로그인 여부에 따라 클라이언트 라우팅을 제한해야합니다! 기존 Link 컴포넌트를 사용하면 to를 props로 전달하여 to의 경로로 라우팅을 수행합니다. 이를 활용하려면 로그인 여부에 따라 to props를 상위 컴포넌트에서 바꿔주어서 리렌더링을 수행해야합니다! 

리렌더링할 필요없이 클릭 이벤트에 따라서 이벤트 리스너에서 라우팅 로직만 컨트롤하는 것이 효율적이지 않을까 생각해서 react-router에서 제공하는 history props를 활용해서 직접 라우팅 로직에 history를 추가하는 식으로 구현했습니다 :)

jinhyukoo reviewed (approved):
@wooojini 고생 많으셨습니다!
코드 유심히 읽어봤는데 딱히 문제가 없어 보이는데 그런 현상을 보인다는게 이상하네요...
혹시나 debounce와 연관이 있을까 생각을 해봐도 클릭을 했을 때 이벤트가 발생하는거라 debouncing 이후 리렌더링이 돼도 아무 문제가 없어야 정상인데말이죠... 이벤트 리스너도 onClick에 잘 등록되어 있구요.

제 기억이 맞다면 debounce 자체가 콜백함수를 받아서 실행하는 형태로 되어 있을 것 같은데 debounce의 콜백으로 click listener를 넘겨주고 있는게 아닌가 한 번 확인해보시는게 어떤가요?

우진님이 Input hook 만드실 때 debounce도 설정할 수 있게 하셨다고 한 거 보면 뭔가 input에 변화가 감지되어서 자동으로 등록된 콜백 함수가 실행되고 있는 것 같아요.

혹시 그런 형태로 되어 있다면 중복확인 같은 경우 입력이 모두 끝난 후 항상 버튼을 누르게 되어 있으니 debounce 설정을 하지 않고 지금 구조 그대로 가져가면 오류가 생기지 않을까 하는 생각입니다!

코드 깔끔하게 잘 작성해주셔서 감사해요 👍

wooojini commented:
@jinhyukoo 리뷰 감사합니당! 리뷰해주신 내용에 대한 제 의견은 아래와 같습니다!

**1) debounce 자체가 콜백함수를 받아서 실행하는 형태로 되어 있을 것 같은데 debounce의 콜백으로 click listener를 넘겨주고 있는게 아닌가?**
  - 그렇지는 않습니다! useInputForm hook을 실행하면 debounce를 적용한 onInputChange를 리턴해줍니다. 이 onChange 함수를 TextField로 전달해주는 주는 것이지요.
  -  onInputChange는 useInputForm hook 내부에서 debounce 함수로 감싸주어 반환된 함수로, input 값 state 변경 및 유효성 검사를 위한 로직만 실행합니다!

**2)  Input hook 만드실 때 debounce도 설정할 수 있게 하셨다고 한 거 보면 뭔가 input에 변화가 감지되어서 자동으로 등록된 콜백 함수가 실행되고 있는 것 같아요**
   - useInputForm hook 내부에서 debouncing을 적용하여 실행하는 함수는 위에 설명된 내용과 같습니다! 

**3) 혹시 그런 형태로 되어 있다면 중복확인 같은 경우 입력이 모두 끝난 후 항상 버튼을 누르게 되어 있으니 debounce 설정을 하지 않고 지금 구조 그대로 가져가면 오류가 생기지 않을까 하는 생각입니다!**
   - 이 부분은 좀 더 구체적으로 설명 괜찮으실까요?? 현재 입력이 모두 끝난 후, 중복체크 버튼을 누르게 되어있는데 debounce 설정을 하지 않는다는 것이 중복체크 쿼리를 날리는 부분이 debounce 로직에 추가되어야 문제가 없을 것 같다는 의미일가요?
   - 만약 위와 같은 의미라면 문제는 없습니다! input에 값을 입력하면 debouce가 실행되고, input state 변경 및 유효성 검사 후, 컴포넌트가 리렌더링됩니다!
   - useLazyQuery는 반환 값으로 쿼리를 수행하는 함수를 리턴하는데 이 함수는 매번 variable값을 가진 새로운 함수로 반환됩니다! 쿼리를 실행할 때는 변경된 input state를 variable로 가지고 실행하기 때문에 문제는 없어보입니당!

wooojini committed:
fix: 회원가입 폼 리렌더링 시 중복체크 API 자동 호출 버그 수정

- useLazyQuery의 옵션으로 variables를 전달하여 이메일이나 닉네임 상태 값 변경 시 쿼리가 업데이트 되어서 자동으로 중복체크 쿼리가 요청되는 버그 발생
- 쿼리의 variables 옵션을 쿼리 함수 호출시 옵션으로 전달하여 버그 수정

wooojini committed:
feat: 이메일, 닉네임 중복체크 버튼 비활성화 로직 추가

- 의도하지 않은 동작을 예방하기 위해 이메일, 닉네임 입력 값이 없거나 유효성 검사를 통과하지 않으면 중복체크
  버튼이 비활성화 되도록 로직 추가

wooojini commented:
@jinhyukoo 

위에서 언급했던 useLazyQuery가 자동으로 실행되던 이슈를 해결했습니다 :)

- 정확한 이슈는 디바운싱에 따라 쿼리가 실행된 것이 아닌 컴포넌트가 리렌더링 될 때마다 쿼리가 실행되는 이슈였습니다.
- useLazyQuery 공식 문서를 보면 쿼리가 바로 실행되는 것이 아닌 쿼리를 실행할 수 있는 함수를 리턴하고, 이 함수를 필요한 순간에 호출하여 쿼리를 실행할 수 있다고 명세되어있습니다.
- 원인은 useLazyQuery에 옵션으로 넘겨준 variables가 바뀌면 쿼리가 자동으로 업데이트되어 실행되기 때문이었습니다. useLazyQuery를 호출할 때, 이전과 옵션값이 다르면 쿼리를 업데이트하고, 자동으로 재실행을 시켜주는데, 쿼리 실행시점을 컨트롤하기 위해서 useLazyQuery를 사용하는데 이러한 동작을 하는 것이 의문이네요! 공식 문서에 이를 명시하지 않은 것도 의문입니다!
- 문제 해결은 쿼리 함수 호출 시점에서 variables를 인자로 넘겨주어 실행하여 해결했습니다 :)
- 아래는 ApolloClient GitHub 레포에 관련 이슈입니다
   - 공식문서에 관련내용이 없고, useLazyQuery가 이런 동작을 하는 것은 문제라고 언급이 되어있네요!
   - https://github.com/apollographql/apollo-client/issues/5912

wooojini commented:
@jinhyukoo 

쿼리 실행 방법이 변경되면서 회원가입 폼 로직 수정이 필요한데, 이 부분은 머지 후 작업하겠습니다 :)

wooojini committed:
Merge branch 'develop' into feat/119-login
